compiler/src/iree/compiler/Dialect/Stream/Transforms/UnifyEncodingForGlobals.cpp

```diff
@@ -270,9 +271,102 @@ GlobalEncodingAnalyzer::traceToSourceGlobal(Value value) {
   return foundSourceGlobal;
 }
 
-//===----------------------------------------------------------------------===//
-// Pass Implementation
-//===----------------------------------------------------------------------===//
+// Maps a tensor op to a set of (operand index -> new encoding).
+using OperandEncodingUpdates = llvm::DenseMap<unsigned, Attribute>;
+using TensorEncodingUpdates =
+    llvm::DenseMap<Operation *, OperandEncodingUpdates>;
+
+// Updates encoding attributes for a TensorDispatchOp.
+static void
+updateTensorDispatchOp(TensorDispatchOp dispatchOp,
+                       const OperandEncodingUpdates &operandUpdates) {
+  // Update operand encodings.
+  // The operand_encodings attribute has the same length as getMixedOperands().
+  // For non-affinity types (e.g., index), the encoding is just the type.
+  // For affinity types, the encoding is a RankedTensorType with encoding attr.
+  SmallVector<Attribute> newOperandEncodings;
+  for (auto [idx, operand, typeAttr] :
+       llvm::enumerate(dispatchOp.getMixedOperands(),
+                       dispatchOp.getOperandEncodings().getValue())) {
+    Type type = cast<TypeAttr>(typeAttr).getValue();
+    if (!isa<IREE::Stream::AffinityTypeInterface>(type)) {
+      newOperandEncodings.push_back(typeAttr);
+      continue;
+    }
+    if (!operandUpdates.contains(idx)) {
+      newOperandEncodings.push_back(typeAttr);
+      continue;
+    }
+    Attribute newEncoding = operandUpdates.lookup(idx);
+    auto tensorType = cast<RankedTensorType>(type);
+    newOperandEncodings.push_back(
+        TypeAttr::get(tensorType.cloneWithEncoding(newEncoding)));
+    LDBG() << "  Updated dispatch operand encoding at index " << idx << " to "
+           << newEncoding;
+  }
+  dispatchOp.setOperandEncodingsAttr(
+      ArrayAttr::get(dispatchOp.getContext(), newOperandEncodings));
+}
+
+// Applies all cached encoding updates to tensor ops.
+static void applyTensorEncodingUpdates(TensorEncodingUpdates &updates) {
+  for (auto &[op, operandUpdates] : updates) {
+    // TODO: Handle other TensorPhaseOp ops (TensorFillOp, etc.) via TypeSwitch.
+    if (auto dispatchOp = dyn_cast<TensorDispatchOp>(op)) {
```

Comment:
Should we use `cast<>` here to get an assert if a currently unsupported op appears here?

Reply:
Some ops should be ignored, e.g., `stream.tensor.constant`. I haven't reviewed all the ops, and I will add tests for each op in the future. I think dyn_cast is fine for now, as we remain the encodings unchanged, which seems better than an assertion/crash?

Reply:
Ok, if there are operations that we can ignore, `dyn_cast` is fine.